java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java

```diff
@@ -78,70 +78,55 @@ public void visitNode(Tree tree) {
   }
 
   private static boolean isAnnotated(ClassTree tree) {
-    for (AnnotationTree annotationTree : tree.modifiers().annotations()) {
-      Tree annotationType = annotationTree.annotationType();
-      if (annotationType.is(Tree.Kind.IDENTIFIER) && "FunctionalInterface".equals(((IdentifierTree) annotationType).name())) {
-        return true;
-      }
-    }
-    return false;
+    return tree.symbol().metadata().isAnnotatedWith("java.lang.FunctionalInterface");
   }
 
-  private static boolean hasOneAbstractMethod(ClassTree classTree) {
-    TypeSymbol symbol = classTree.symbol();
-    if (symbol != null) {
-      List<Type> types = symbol.interfaces();
-      for (Type type : types) {
-        if (!type.symbol().memberSymbols().isEmpty()) {
-          return false;
-        }
-      }
-    }
-    int methods = 0;
-    for (Tree member : classTree.members()) {
-      boolean isMethod = member.is(Tree.Kind.METHOD);
-      if (!isMethod) {
-        return false;
-      }
-      if (isNotObjectMethod((MethodTree) member) && isNonStaticNonDefaultMethod(member)) {
-        methods++;
-      }
-    }
-    return methods == 1;
+  private static boolean hasOneAbstractMethod(Symbol.TypeSymbol symbol) {
+    return numberOfAbstractMethod(symbol) == 1 && noAbstractMethodInParentInterfaces(symbol.interfaces());
   }
 
-  private static boolean isNotObjectMethod(MethodTree method) {
-    ImmutableCollection<List<String>> methods = OBJECT_METHODS.get(method.simpleName().name());
-    if (methods != null) {
-      for (List<String> types : methods) {
-        if (sameParameters(method.parameters(), types)) {
-          return false;
-        }
-      }
-    }
-    return true;
+  private static boolean noAbstractMethodInParentInterfaces(List<Type> interfaces) {
+    return interfaces.stream()
+      .map(Type::symbol)
+      .noneMatch(symbol -> numberOfAbstractMethod(symbol) > 0 || !noAbstractMethodInParentInterfaces(symbol.interfaces()));
   }
 
-  private static boolean sameParameters(List<VariableTree> provided, List<String> expectedTypes) {
-    List<String> args = Lists.newArrayList(expectedTypes);
-    if (provided.size() == args.size()) {
-      for (VariableTree var : provided) {
-        args.remove(var.type().symbolType().name());
-      }
-      if (args.isEmpty()) {
-        return true;
-      }
+  private static long numberOfAbstractMethod(Symbol symbol) {
+    if (!symbol.isTypeSymbol()) {
+      // unknown interface
+      return Integer.MAX_VALUE;
     }
-    return false;
+    Symbol.TypeSymbol interfaceSymbol = (Symbol.TypeSymbol) symbol;
+    return interfaceSymbol.memberSymbols().stream()
+      .filter(member -> member.isMethodSymbol())
+      .filter(member -> {
+        Symbol.MethodSymbol methodSymbol = (Symbol.MethodSymbol) member;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 11. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)